File: src/ts/reference/core/deployment/paths.ts
import path from "node:path";
import { getLoomPaths } from "../state/paths";

export type DeploymentPaths = {
  specsDir: string;
  stateDir: string;
  logsDir: string;
  repairsDir: string;
  evidenceDir: string;
  specFile: string;
  stateFile: string;
  activeOperationFile: string;
  staleOperationFile: string;
  logFile: string;
  repairFile: string;
  failureFile: string;
  codeEvidenceFile: string;
  generatedDir: string;
  dockerfileFile: string;
  composeFile: string;
  dockerignoreFile: string;
};

export function getDeploymentPaths(projectRoot: string): DeploymentPaths {
  const paths = getLoomPaths(projectRoot);
  const deploymentDir = path.join(paths.loomDir, "deployment");
  const specsDir = path.join(deploymentDir, "specs");
  const stateDir = path.join(deploymentDir, "state");
  const logsDir = path.join(deploymentDir, "logs");
  const repairsDir = path.join(deploymentDir, "repairs");
  const evidenceDir = path.join(deploymentDir, "evidence");
  const generatedDir = path.join(specsDir, "generated");

  return {
    specsDir,
    stateDir,
    logsDir,
    repairsDir,
    evidenceDir,
    specFile: path.join(specsDir, "local.json"),
    stateFile: path.join(stateDir, "local.json"),
    activeOperationFile: path.join(stateDir, "active-operation.json"),
    staleOperationFile: path.join(stateDir, "last-stale-operation.json"),
    logFile: path.join(logsDir, "local.log"),
    repairFile: path.join(stateDir, "repair-request.json"),
    failureFile: path.join(stateDir, "latest-failure.json"),
    codeEvidenceFile: path.join(evidenceDir, "latest-code-evidence.json"),
    generatedDir,
    dockerfileFile: path.join(generatedDir, "Dockerfile"),
    composeFile: path.join(generatedDir, "compose.yaml"),
    dockerignoreFile: path.join(generatedDir, "Dockerfile.dockerignore"),
  };
}

export function getDeploymentRepairPaths(projectRoot: string, repairId: string): {
  repairDir: string;
  requestFile: string;
  taskExecutionRequestFile: string;
  resultFile: string;
} {
  const paths = getDeploymentPaths(projectRoot);
  const repairDir = path.join(paths.repairsDir, repairId);
  return {
    repairDir,
    requestFile: path.join(repairDir, "request.json"),
    taskExecutionRequestFile: path.join(repairDir, "task-execution-request.json"),
    resultFile: path.join(repairDir, "result.json"),
  };
}

File: src/rust/deploy/paths.rs
use std::path::{Path, PathBuf};

use state::paths::LOOM_DIR;

#[derive(Debug, Clone)]
pub struct DeploymentPaths {
    pub specs_dir: PathBuf,
    pub state_dir: PathBuf,
    pub logs_dir: PathBuf,
    pub repairs_dir: PathBuf,
    pub evidence_dir: PathBuf,
    pub spec_file: PathBuf,
    pub state_file: PathBuf,
    pub active_operation_file: PathBuf,
    pub stale_operation_file: PathBuf,
    pub log_file: PathBuf,
    pub repair_action_file: PathBuf,
    pub failure_file: PathBuf,
    pub code_evidence_file: PathBuf,
    pub generated_dir: PathBuf,
    pub compose_file: PathBuf,
    pub dockerignore_file: PathBuf,
}

pub fn deployment_paths(project_root: &Path) -> DeploymentPaths {
    let deployment_dir = project_root.join(LOOM_DIR).join("deployment");
    let specs_dir = deployment_dir.join("specs");
    let state_dir = deployment_dir.join("state");
    let logs_dir = deployment_dir.join("logs");
    let repairs_dir = deployment_dir.join("repairs");
    let evidence_dir = deployment_dir.join("evidence");
    let generated_dir = specs_dir.join("generated");
    DeploymentPaths {
        spec_file: specs_dir.join("local.json"),
        state_file: state_dir.join("local.json"),
        active_operation_file: state_dir.join("active-operation.json"),
        stale_operation_file: state_dir.join("last-stale-operation.json"),
        log_file: logs_dir.join("local.log"),
        repair_action_file: state_dir.join("repair-action.json"),
        failure_file: state_dir.join("latest-failure.json"),
        code_evidence_file: evidence_dir.join("latest-code-evidence.json"),
        compose_file: generated_dir.join("compose.yaml"),
        dockerignore_file: generated_dir.join("Dockerfile.dockerignore"),
        specs_dir,
        state_dir,
        logs_dir,
        repairs_dir,
        evidence_dir,
        generated_dir,
    }
}

pub fn dockerfile_path(project_root: &Path, service_id: &str) -> PathBuf {
    deployment_paths(project_root)
        .generated_dir
        .join(format!("Dockerfile.{service_id}"))
}

pub fn dockerfile_ignore_path(project_root: &Path, service_id: &str) -> PathBuf {
    deployment_paths(project_root)
        .generated_dir
        .join(format!("Dockerfile.{service_id}.dockerignore"))
}

pub fn nginx_config_path(project_root: &Path, service_id: &str) -> PathBuf {
    deployment_paths(project_root)
        .generated_dir
        .join(format!("nginx.{service_id}.conf"))
}

pub fn deploy_execution_repair_result_file(project_root: &Path, request_id: &str) -> PathBuf {
    project_root
        .join(LOOM_DIR)
        .join("agent-writable")
        .join(request_id)
        .join("deploy-execution-repair-result.json")
}

pub fn deploy_execution_repair_action_file(project_root: &Path, request_id: &str) -> PathBuf {
    deployment_paths(project_root)
        .repairs_dir
        .join(request_id)
        .join("request.json")
}
